Introduce channels for staging partition write and commit

shixuan-fan opened the pull request:
Extracted from #12529 ; part of #12124

We removed hive session property `staging_file_writing_enabled`, instead, we let the engine decide when to write and commit staging partitions. The engine will use `PageSink` as the channel to tell the connector to write staging partitions, and use `Metadata` to commit the partitions. In this PR we introduced these two channels, and they will be used in #12529.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1143,13 +1142,6 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
-            stagingFileCommitter.commitFiles(session, handle.getSchemaName(), handle.getTableName(), partitionUpdates);
-        }
-        else {
-            verifyNoStagingFilesWritten(partitionUpdates);
```
nit: do we want to keep this verification for now? If no - should we remove the `verifyNoStagingFilesWritten` method itself?

shixuan-fan replied:
Good point! We should still keep this verification until recoverable execution kicks in. Let me change that.

shixuan-fan replied:
As per offline discussion, we will remove all the checks here because #12529 will be merged shortly which will again remove the check.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -270,6 +271,8 @@
 import static org.testng.Assert.assertTrue;
 import static org.testng.Assert.fail;
 
+();
```
Most likely unintended change

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/PageSinkProperties.java:
```diff
@@ -0,0 +1,57 @@
+        return DEFAULT_PAGE_SINK_PROPERTY;
+    }
+
+    public static Builder builder()
```
nit: move it below the `isPartitionCommitRequired` method

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/PageSinkProperties.java:
```diff
@@ -0,0 +1,57 @@
+
+public class PageSinkProperties
+{
+    public static final PageSinkProperties DEFAULT_PAGE_SINK_PROPERTY = PageSinkProperties.builder().build();
```
DEFAULT_PAGE_SINK_PROPERTIES

arhimondr commented on presto-main/src/main/java/com/facebook/presto/split/PageSinkManager.java:
```diff
@@ -46,19 +47,19 @@ public void removeConnectorPageSinkProvider(ConnectorId connectorId)
 
     @Override
-    public ConnectorPageSink createPageSink(Session session, OutputTableHandle tableHandle)
+    public ConnectorPageSink createPageSink(Session session, OutputTableHandle tableHandle, PageSinkProperties property)
```
`properties`

shixuan-fan replied:
I'll name them `pageSinkProperties` to be consistent with others.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/split/PageSinkManager.java:
```diff
@@ -46,19 +47,19 @@ public void removeConnectorPageSinkProvider(ConnectorId connectorId)
 
     @Override
-    public ConnectorPageSink createPageSink(Session session, InsertTableHandle tableHandle)
+    public ConnectorPageSink createPageSink(Session session, InsertTableHandle tableHandle, PageSinkProperties property)
```
`properties`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/split/PageSinkProvider.java:
```diff
@@ -17,10 +17,11 @@
 public interface PageSinkProvider
 {
-    ConnectorPageSink createPageSink(Session session, OutputTableHandle tableHandle);
+    ConnectorPageSink createPageSink(Session session, OutputTableHandle tableHandle, PageSinkProperties property);
```
`properties`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/split/PageSinkProvider.java:
```diff
@@ -17,10 +17,11 @@
+    ConnectorPageSink createPageSink(Session session, OutputTableHandle tableHandle, PageSinkProperties property);
 
-    ConnectorPageSink createPageSink(Session session, InsertTableHandle tableHandle);
+    ConnectorPageSink createPageSink(Session session, InsertTableHandle tableHandle, PageSinkProperties property);
```
`properties`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1143,13 +1142,6 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
         List<PartitionUpdate> partitionUpdates = getPartitionUpdates(fragments);
 
-        if (isWritingStagingFilesEnabled(session)) {
-            stagingFileCommitter.commitFiles(session, handle.getSchemaName(), handle.getTableName(), partitionUpdates);
```
I wonder if the SPI change that exposes this method to the engine should also be included in this PR. This will create better visibility of what is tried to be done here.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -2153,6 +2153,20 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
     }
 
+    @Override
+    public void commitPartitionForCreate(ConnectorSession session, ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments)
```
Let's use plural (`commitPartitions`) here, as this method will be used to commit multiple partitions (buckets) at a time.

arhimondr replied:
We should rename it to `commitPartitionForCreateTable` to be consistent with the `beginCreateTable`  and the `finishCreateTable`

arhimondr replied:
Also i think we may consider calling both methods simply `commitPartitions`. The ovearloading of the `ConnectorOutputTableHandle` / `ConnectorInsertTableHandle` should make it clean why there are two methods.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -2153,6 +2153,20 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
+    }
+
+    @Override
+    public void commitPartitionForInsert(ConnectorSession session, ConnectorInsertTableHandle tableHandle, Collection<Slice> fragments)
```
Since the `testInsertAndCreateTableWithWritingStagingFileEnabled` test has been removed, i would suggest adding an integration test to the `AbstractTestHiveClient` that creates a PageSink with a partition commit enabled, and then commits the results with the new SPI.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -3201,7 +3226,7 @@ private void doCreateView(SchemaTableName viewName, boolean replace)
     }
 
-    protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storageFormat)
+    protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storageFormat, PageSinkProperties pageSinkProperties)
```
nit: passing `PageSinkProperties pageSinkProperties` seems to be unnecessary here. How about a simple boolean flag, `boolean partitionCommitRequired`

shixuan-fan replied:
We need to pass `PageSinkProperties` for `pageSinkProvider.createPageSink`. Are you suggesting that we use a boolean flag and construct `PageSinkProperties` in the `doFoo` method?

arhimondr replied:
> We need to pass PageSinkProperties for pageSinkProvider.createPageSink

I missed that. Nevermind.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -2584,6 +2584,31 @@ public void testInsertUnsupportedWriteType()
     }
 
+    @Test
+    public void testStagingPartition()
```
nit: `testCommitPartitions`?

shixuan-fan replied:
I removed this test and added tests in `testCreate`, `testInsert`, `testInsertIntoExistingPartition` and `testInsertIntoNewPartition`.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -3329,7 +3358,7 @@ protected void assertEmptyTableStatistics(SchemaTableName tableName, Transaction
     }
 
-    private void doInsert(HiveStorageFormat storageFormat, SchemaTableName tableName)
+    private void doInsert(HiveStorageFormat storageFormat, SchemaTableName tableName, PageSinkProperties pageSinkProperties)
```
nit: Ditto about the `partitionCommitRequired`

arhimondr commented:
> Introduce channels for staging partition write and commit

The PR name seems to be incorrect. This PR introduces the `commitPartitions` SPI.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -3216,10 +3241,14 @@ protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storag
             Collection<Slice> fragments = getFutureValue(sink.finish());
 
+            if (pageSinkProperties.isPartitionCommitRequired()) {
+                metadata.commitPartitions(session, outputHandle, fragments);
```
Let's also decode the `fragments` and check that the writePath and the targetPath are different if commit is required

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -3380,10 +3409,13 @@ private void doInsert(HiveStorageFormat storageFormat, SchemaTableName tableName
             sink.appendPage(CREATE_TABLE_DATA.toPage());
             Collection<Slice> fragments = getFutureValue(sink.finish());
+            if (pageSinkProperties.isPartitionCommitRequired()) {
+                metadata.commitPartitions(session, insertTableHandle, fragments);
```
ditto

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -3380,10 +3409,13 @@ private void doInsert(HiveStorageFormat storageFormat, SchemaTableName tableName
             Collection<Slice> fragments = getFutureValue(sink.finish());
+            if (pageSinkProperties.isPartitionCommitRequired()) {
+                metadata.commitPartitions(session, insertTableHandle, fragments);
+            }
```
Let's also do similar tests for `doInsertIntoNewPartition`, `testInsertIntoExistingPartition`

shixuan-fan commented:
Comment addressed. As discussed offline, this PR introduces two SPI changes: `commitPartitions` and `PageSinkProperties`, so the current title is okay since it covers both ends.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/PageSinkProperties.java:
```diff
@@ -0,0 +1,57 @@
+
+public class PageSinkProperties
+{
+    public static final PageSinkProperties DEFAULT_PAGE_SINK_PROPERTIES = PageSinkProperties.builder().build();
```
nit: use `private` ?

wenleix commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -2153,6 +2153,20 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
     }
 
+    @Override
+    public void commitPartitions(ConnectorSession session, ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments)
```
nit: `commitPartition` since it's one partition at a time commit.

arhimondr replied:
We use the word "partition" here in terms of the Presto "partition", that in Hive maps to "bucket". In Hive 1 bucket = 1 "partition"  = 1 file = 1 fragment (partition update).

shixuan-fan replied:
`commitFiles` could happen because for each presto partition, we might need to write to multiple hive partitions, and thus multiple directories. In each directory there would be one file. One fragment in Hive (`PartitionUpdate`) is one directory (as you could see, there is one path and a list of file names).

arhimondr replied:
I see. Thanks for the explanation.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -629,4 +629,20 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
     }
+
+    /**
+     * Commits partition for table creation.
```
Maybe also explain that for connector supports partition commit, engine can do recoverable grouped execution?

shixuan-fan replied:
"for connector supports partition commit, engine can do recoverable grouped execution" is not exactly correct because we have other limit as well (I wish we could just claim that :D). How about "To enable recoverable grouped execution, it is required that output connector supports partition commit"?

wenleix commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -3759,6 +3791,16 @@ private void doInsertIntoExistingPartition(HiveStorageFormat storageFormat, Sche
         }
     }
 
+    private static void assertPartitionCommitFragments(Collection<Slice> fragments)
```
nit: `assertValidPartitionCommitFragments` and add a one line comment?

shixuan-fan commented:
Thanks for reviewing! Two things worth mentioning:

- I realized that we should add the corresponding connector capability for Hive and I forgot to do so in #12677 (rewindable split source). So I added a small commit to amend that.
- To make `commitPartition` API semantically clear, I added `partitionId` in the API, which corresponds to `Lifespan::groupId` from engine's perspective. Though for Hive it is not required, but from an API's perspective it is good to have and might be required for other connectors.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -629,4 +629,22 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
+    /**
+     * Commits partition for table creation.
+     * To enable recoverable grouped execution, it is required that output connector supports partition commit.
+     */
```
don't forget this. ditto for another commit partition.
```
@apiNote This method is unstable and subject to change in the future.
```

wenleix commented on presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java:
```diff
@@ -398,6 +398,16 @@
     List<GrantInfo> listTablePrivileges(Session session, QualifiedTablePrefix prefix);
 
+    /**
+     * Commits partition for table creation.
```
api note about unstable and subject to change :)
